File: practices/28-scatter-gather/src/python/research_worker.py
"""
Practice 28: Research Worker Agent

A worker agent that researches a single subtopic within a codebase.
Workers operate independently and produce structured findings.
"""

import json
import re
from claude_agent_sdk import Agent


class ResearchWorker:
    """Worker agent that researches a single subtopic."""

    def __init__(self, worker_id: int = 0):
        self.worker_id = worker_id
        self.agent = Agent()

    def research(self, subtopic: str, codebase_path: str) -> dict:
        """
        Research a single subtopic within the codebase.

        Args:
            subtopic: The focused research subtopic
            codebase_path: Path to the codebase to research

        Returns:
            Dict with: subtopic, findings, key_files, confidence, recommendations
        """
        result = self.agent.query(
            prompt=(
                f"You are Research Worker #{self.worker_id}.\n\n"
                f"## Your Research Task\n"
                f"Research this subtopic within the codebase at {codebase_path}:\n\n"
                f"**Subtopic:** {subtopic}\n\n"
                f"## Instructions\n"
                f"1. Search the codebase thoroughly for information related to this subtopic\n"
                f"2. Read relevant files to understand patterns and approaches\n"
                f"3. Note specific file paths and line numbers for key findings\n"
                f"4. Rate your confidence in your findings\n\n"
                f"## Output Format\n"
                f"Return your findings in this exact JSON format:\n"
                f'{{\n'
                f'  "subtopic": "{subtopic}",\n'
                f'  "findings": "Detailed description of what you found",\n'
                f'  "key_files": ["file1.ts:10-20", "file2.py:30-40"],\n'
                f'  "confidence": "low|medium|high",\n'
                f'  "recommendations": ["recommendation 1", "recommendation 2"],\n'
                f'  "evidence_count": 0\n'
                f'}}\n\n'
                f"Be thorough but focused. Only research this specific subtopic."
            ),
            allowed_tools=["Read", "Glob", "Grep"],
        )

        # Try to parse JSON from the response
        try:
            json_match = re.search(r"\{.*\}", result.text, re.DOTALL)
            if json_match:
                parsed = json.loads(json_match.group())
                return {
                    "subtopic": subtopic,
                    "findings": parsed.get("findings", result.text),
                    "key_files": parsed.get("key_files", []),
                    "confidence": parsed.get("confidence", "medium"),
                    "recommendations": parsed.get("recommendations", []),
                    "evidence_count": parsed.get("evidence_count", 0),
                    "raw_text": result.text,
                    "tokens_used": getattr(result, "usage", None),
                }
        except (json.JSONDecodeError, AttributeError):
            pass

        # Fallback: return the raw text as findings
        return {
            "subtopic": subtopic,
            "findings": result.text,
            "key_files": [],
            "confidence": "medium",
            "recommendations": [],
            "evidence_count": 0,
            "raw_text": result.text,
            "tokens_used": getattr(result, "usage", None),
        }

File: practices/28-scatter-gather/src/typescript/research-worker.ts
/**
 * Practice 28: Research Worker Agent (TypeScript)
 *
 * A worker agent that researches a single subtopic within a codebase.
 * Workers operate independently and produce structured findings.
 */

import { Agent } from "@anthropic-ai/claude-agent-sdk";
import { WorkerFinding } from "./research-lead";

export class ResearchWorker {
  private workerId: number;
  private agent: Agent;

  constructor(workerId: number = 0) {
    this.workerId = workerId;
    this.agent = new Agent();
  }

  /**
   * Research a single subtopic within the codebase.
   */
  async research(
    subtopic: string,
    codebasePath: string
  ): Promise<WorkerFinding> {
    const result = await this.agent.query({
      prompt:
        `You are Research Worker #${this.workerId}.\n\n` +
        `## Your Research Task\n` +
        `Research this subtopic within the codebase at ${codebasePath}:\n\n` +
        `**Subtopic:** ${subtopic}\n\n` +
        `## Instructions\n` +
        `1. Search the codebase thoroughly for information related to this subtopic\n` +
        `2. Read relevant files to understand patterns and approaches\n` +
        `3. Note specific file paths and line numbers for key findings\n` +
        `4. Rate your confidence in your findings\n\n` +
        `## Output Format\n` +
        `Return your findings in this exact JSON format:\n` +
        `{\n` +
        `  "subtopic": "${subtopic}",\n` +
        `  "findings": "Detailed description of what you found",\n` +
        `  "key_files": ["file1.ts:10-20", "file2.py:30-40"],\n` +
        `  "confidence": "low|medium|high",\n` +
        `  "recommendations": ["recommendation 1", "recommendation 2"],\n` +
        `  "evidence_count": 0\n` +
        `}\n\n` +
        `Be thorough but focused. Only research this specific subtopic.`,
      allowedTools: ["Read", "Glob", "Grep"],
    });

    try {
      const jsonMatch = result.text.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        const parsed = JSON.parse(jsonMatch[0]);
        return {
          subtopic,
          findings: parsed.findings || result.text,
          keyFiles: parsed.key_files || [],
          confidence: parsed.confidence || "medium",
          recommendations: parsed.recommendations || [],
          evidenceCount: parsed.evidence_count || 0,
          rawText: result.text,
        };
      }
    } catch {
      // Fall through to default
    }

    return {
      subtopic,
      findings: result.text,
      keyFiles: [],
      confidence: "medium",
      recommendations: [],
      evidenceCount: 0,
      rawText: result.text,
    };
  }
}
